File: 0383-ransom-note/0383-ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> letterCount;
        
        for (char ch: magazine) {
            letterCount[ch]++;
        }
        
        for (char ch: ransomNote) {
            if (letterCount[ch] == 0) {
                return false;
            }
            letterCount[ch]--;
        }
        return true;
    }
};

File: 0383-ransom-note/0383-ransom-note.rs
use std::collections::HashMap;

impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut letter_count = HashMap::new();

        for char in magazine.chars() {
            *letter_count.entry(char).or_insert(0) += 1;
        }

        for char in ransom_note.chars() {
            let count = letter_count.entry(char).or_insert(0);

            if *count == 0 {
                return false;
            }
            *count -= 1;
        }

        true
    }
}